wp-includes/rest-api/auth/class-wp-rest-key-pair.php

```diff
@@ -0,0 +1,798 @@
+	 */
+	public function show_user_profile( WP_User $user ) {
+		wp_enqueue_style( 'key-pair-css', JWT_AUTH_PLUGIN_URL . '/wp-admin/css/key-pair.css', array(), JWT_AUTH_VERSION );
+		wp_enqueue_script( 'key-pair-js', JWT_AUTH_PLUGIN_URL . '/wp-admin/js/key-pair.js', array(), JWT_AUTH_VERSION, true );
+		wp_localize_script(
+			'key-pair-js',
+			'keyPair',
+			array(
+				'nonce'   => wp_create_nonce( 'wp_rest' ),
+				'root'    => self::get_rest_uri(),
+				'token'   => WP_REST_Token::get_rest_uri(),
+				'user_id' => $user->ID,
+				'text'    => array(
+					/* translators: %s: key-pair name */
+					'confirm_one' => esc_html__( 'Revoke the %s key-pair? This action cannot be undone.', 'jwt-auth' ),
+					'confirm_all' => esc_html__( 'Revoke all key-pairs? This action cannot be undone.', 'jwt-auth' ),
+				),
+			)
+		);
+		$this->show_key_pair_section( $user );
+		$this->template_new_key_pair();
+		$this->template_new_token_key_pair();
+		$this->template_key_pair_row();
+	}
+
+	/**
+	 * Fires after the user's password is reset.
+	 *
+	 * @param WP_User $user The user.
+	 */
+	public function after_password_reset( WP_User $user ) {
+		if ( 'after_password_reset' !== current_filter() ) {
+			return;
+		}
+
+		$keypairs = $this->get_user_key_pairs( $user->ID );
+		if ( ! empty( $keypairs ) ) {
+			$this->set_user_key_pairs( $user->ID, array() );
+		}
+	}
+
+	/**
+	 * Fires after the user's password is reset.
+	 *
+	 * @param int $user_id The user ID.
+	 */
+	public function profile_update( $user_id ) {
+		if ( 'profile_update' !== current_filter() ) {
+			return;
+		}
+
+		if ( isset( $_POST['pass1'] ) && ! empty( $_POST['pass1'] ) ) { // phpcs:ignore
+			$keypairs = $this->get_user_key_pairs( $user_id );
+			if ( ! empty( $keypairs ) ) {
+				$this->set_user_key_pairs( $user_id, array() );
+			}
+		}
+	}
+
+	/**
+	 * Filters `rest_authentication_require_token` to exclude the key-pair endpoint,
+	 *
+	 * @param bool   $require_token Whether a token is required.
+	 * @param string $request_uri The URI which was given by the server.
+	 * @param string $request_method Which request method was used to access the server.
+	 *
+	 * @return bool
+	 */
+	public function require_token( $require_token, $request_uri, $request_method ) {
+
+		// Don't require token authentication to manage key-pairs.
+		if ( ( 'POST' === $request_method || 'DELETE' === $request_method ) && false !== strpos( $request_uri, self::get_rest_uri() ) ) {
+			$require_token = false;
+		}
+
+		return $require_token;
+	}
+
+	/**
+	 * Authenticate the key-pair if API key and API secret is provided and return the user.
+	 *
+	 * If not authenticated, send back the original $user value to allow other authentication
+	 * methods to attempt authentication.
+	 *
+	 * @filter rest_authentication_user
+	 *
+	 * @param mixed           $user    The user that's being authenticated.
+	 * @param WP_REST_Request $request The REST request object.
+	 *
+	 * @return bool|object|mixed
```

Comment:
Suggestion: Add the objects returned: `WP_User` and `WP_Error` so that I don't have to read the code/method to know which objects are returned.

Reply:
Technically if someone has filtered `rest_authentication_user` it could be any kind of object. If `$user` is not equal to `false` the method returns the value so yes the intention is that it will return a `WP_User` but that's not a guarantee and will work perfectly with any type of object with the correct properties. If someone wanted to build their own decoupled user system that's possible by filtering `rest_authentication_user`. I'll update the docblock to call this out.

Reply:
K! Thank you! Makes sense!